File: taskm-sdk-py/taskm_sdk/logger.py
"""
Logger utility for HS-TASKM SDK.

This module provides a simple logging utility that automatically configures
logging based on environment variables and provides convenient logging methods.
"""

import os
import sys
import logging
import json
from datetime import datetime
from typing import Any, Optional, Dict


class Logger:
    """
    Simple logger for TaskM strategy execution.

    This logger provides convenient logging methods with automatic configuration
    from environment variables. It supports both plain text and JSON logging.

    Environment variables:
    - LOG_LEVEL: Logging level (DEBUG, INFO, WARNING, ERROR, default: INFO)
    - LOG_FORMAT: Log format (text, json, default: text)
    - LOG_TIME_FORMAT: Time format for text logs (default: %Y-%m-%d %H:%M:%S)

    Example:
        >>> from taskm_sdk import Logger
        >>> logger = Logger("MyStrategy")
        >>> logger.info("Strategy started")
        >>> logger.debug("Processing data", symbol="BTC/USDT")
        >>> logger.error("Failed to execute", error=str(e))
    """

    # Log level mapping
    LEVELS = {
        'DEBUG': logging.DEBUG,
        'INFO': logging.INFO,
        'WARNING': logging.WARNING,
        'ERROR': logging.ERROR,
        'CRITICAL': logging.CRITICAL,
    }

    def __init__(
        self,
        name: str,
        level: Optional[str] = None,
        format_type: Optional[str] = None,
        handler: Optional[logging.Handler] = None
    ):
        """
        Initialize the logger.

        Args:
            name: Logger name (usually strategy or component name)
            level: Log level (DEBUG, INFO, WARNING, ERROR). If None, reads from LOG_LEVEL env var
            format_type: Log format type ('text' or 'json'). If None, reads from LOG_FORMAT env var
            handler: Custom logging handler. If None, logs to stdout
        """
        self.name = name
        self.logger = logging.getLogger(name)
        self.logger.setLevel(logging.DEBUG)

        # Determine log level from parameter or environment
        if level is None:
            level = os.getenv('LOG_LEVEL', 'INFO').upper()
        self.level = self.LEVELS.get(level, logging.INFO)

        # Determine format type from parameter or environment
        if format_type is None:
            format_type = os.getenv('LOG_FORMAT', 'text').lower()
        self.format_type = format_type

        # Clear existing handlers to avoid duplicates
        self.logger.handlers.clear()

        # Create handler
        if handler is None:
            handler = logging.StreamHandler(sys.stdout)

        handler.setLevel(self.level)

        # Set formatter
        if self.format_type == 'json':
            formatter = JsonFormatter(name)
        else:
            time_format = os.getenv('LOG_TIME_FORMAT', '%Y-%m-%d %H:%M:%S')
            formatter = logging.Formatter(
                fmt='%(asctime)s - %(name)s - %(levelname)s - %(message)s',
                datefmt=time_format
            )

        handler.setFormatter(formatter)
        self.logger.addHandler(handler)

        # Prevent propagation to avoid duplicate logs
        self.logger.propagate = False

    def debug(self, message: str, **kwargs):
        """Log debug message."""
        extra = kwargs if kwargs else {}
        if self.format_type == 'json':
            self.logger.debug(message, extra={'data': extra})
        else:
            self.logger.debug(f"{message} {extra if extra else ''}".strip())

    def info(self, message: str, **kwargs):
        """Log info message."""
        extra = kwargs if kwargs else {}
        if self.format_type == 'json':
            self.logger.info(message, extra={'data': extra})
        else:
            self.logger.info(f"{message} {extra if extra else ''}".strip())

    def warning(self, message: str, **kwargs):
        """Log warning message."""
        extra = kwargs if kwargs else {}
        if self.format_type == 'json':
            self.logger.warning(message, extra={'data': extra})
        else:
            self.logger.warning(f"{message} {extra if extra else ''}".strip())

    def error(self, message: str, **kwargs):
        """Log error message."""
        extra = kwargs if kwargs else {}
        if self.format_type == 'json':
            self.logger.error(message, extra={'data': extra})
        else:
            self.logger.error(f"{message} {extra if extra else ''}".strip())

    def critical(self, message: str, **kwargs):
        """Log critical message."""
        extra = kwargs if kwargs else {}
        if self.format_type == 'json':
            self.logger.critical(message, extra={'data': extra})
        else:
            self.logger.critical(f"{message} {extra if extra else ''}".strip())


class JsonFormatter(logging.Formatter):
    """JSON formatter for structured logging."""

    def __init__(self, logger_name: str):
        super().__init__()
        self.logger_name = logger_name

    def format(self, record: logging.LogRecord) -> str:
        """Format log record as JSON."""
        log_data = {
            'timestamp': datetime.utcnow().isoformat() + 'Z',
            'logger': self.logger_name,
            'level': record.levelname,
            'message': record.getMessage(),
        }

        # Add extra data if present
        if hasattr(record, 'data') and record.data:
            log_data['data'] = record.data

        # Add exception info if present
        if record.exc_info:
            log_data['exception'] = self.formatException(record.exc_info)

        return json.dumps(log_data)


# Convenience function for quick logger creation
def get_logger(name: str, **kwargs) -> Logger:
    """
    Get or create a logger with the given name.

    Args:
        name: Logger name
        **kwargs: Additional arguments passed to Logger constructor

    Returns:
        Logger instance

    Example:
        >>> logger = get_logger("MyStrategy")
        >>> logger.info("Starting")
    """
    return Logger(name, **kwargs)

File: taskm-sdk-java/src/main/java/com/taskm/sdk/Logger.java
package com.taskm.sdk;

import org.slf4j.LoggerFactory;

import java.util.Map;

/**
 * Logger facade for TaskM SDK.
 *
 * <p>This class provides a simple logging API that wraps SLF4J, the standard logging facade for Java.
 * It supports structured logging with key-value pairs and reads log level from environment variables.</p>
 *
 * <h3>Environment variables:</h3>
 * <ul>
 *   <li>LOG_LEVEL: Logging level (TRACE, DEBUG, INFO, WARN, ERROR, default: INFO)</li>
 *   <li>LOG_FORMAT: Log format (text, json, default: text) - Note: JSON format requires Logback config</li>
 * </ul>
 *
 * <h3>Example usage:</h3>
 * <pre>{@code
 * import com.taskm.sdk.Logger;
 *
 * Logger logger = new Logger("MyStrategy");
 * logger.info("Strategy started");
 * logger.debug("Processing data", "symbol", "BTC/USDT", "price", 50000);
 * logger.error("Failed to execute", "error", exception.getMessage());
 * }</pre>
 *
 * <h3>Logback Configuration (in application):</h3>
 * <p>Your application using this SDK should provide a Logback configuration file
 * (logback.xml or logback-spring.xml) to control log output format and destination.</p>
 *
 * @since 1.0.0
 */
public final class Logger {

    private final org.slf4j.Logger slf4jLogger;

    /**
     * Create a new logger with the given name.
     *
     * @param name the logger name (usually strategy or component name)
     */
    public Logger(String name) {
        if (name == null || name.isEmpty()) {
            throw new IllegalArgumentException("Logger name cannot be null or empty");
        }
        this.slf4jLogger = LoggerFactory.getLogger(name);
    }

    /**
     * Log a debug message.
     *
     * @param message the log message
     */
    public void debug(String message) {
        if (slf4jLogger.isDebugEnabled()) {
            slf4jLogger.debug(message);
        }
    }

    /**
     * Log a debug message with additional data.
     *
     * @param message the log message
     * @param key a data key
     * @param value a data value
     */
    public void debug(String message, String key, Object value) {
        if (slf4jLogger.isDebugEnabled()) {
            slf4jLogger.debug(formatMessage(message, key, value));
        }
    }

    /**
     * Log a debug message with additional data.
     *
     * @param message the log message
     * @param data additional data as key-value pairs
     */
    public void debug(String message, Map<String, Object> data) {
        if (slf4jLogger.isDebugEnabled()) {
            slf4jLogger.debug(formatMessage(message, data));
        }
    }

    /**
     * Log an info message.
     *
     * @param message the log message
     */
    public void info(String message) {
        if (slf4jLogger.isInfoEnabled()) {
            slf4jLogger.info(message);
        }
    }

    /**
     * Log an info message with additional data.
     *
     * @param message the log message
     * @param key a data key
     * @param value a data value
     */
    public void info(String message, String key, Object value) {
        if (slf4jLogger.isInfoEnabled()) {
            slf4jLogger.info(formatMessage(message, key, value));
        }
    }

    /**
     * Log an info message with additional data.
     *
     * @param message the log message
     * @param data additional data as key-value pairs
     */
    public void info(String message, Map<String, Object> data) {
        if (slf4jLogger.isInfoEnabled()) {
            slf4jLogger.info(formatMessage(message, data));
        }
    }

    /**
     * Log a warning message.
     *
     * @param message the log message
     */
    public void warning(String message) {
        if (slf4jLogger.isWarnEnabled()) {
            slf4jLogger.warn(message);
        }
    }

    /**
     * Log a warning message with additional data.
     *
     * @param message the log message
     * @param key a data key
     * @param value a data value
     */
    public void warning(String message, String key, Object value) {
        if (slf4jLogger.isWarnEnabled()) {
            slf4jLogger.warn(formatMessage(message, key, value));
        }
    }

    /**
     * Log a warning message with additional data.
     *
     * @param message the log message
     * @param data additional data as key-value pairs
     */
    public void warning(String message, Map<String, Object> data) {
        if (slf4jLogger.isWarnEnabled()) {
            slf4jLogger.warn(formatMessage(message, data));
        }
    }

    /**
     * Log an error message.
     *
     * @param message the log message
     */
    public void error(String message) {
        slf4jLogger.error(message);
    }

    /**
     * Log an error message with additional data.
     *
     * @param message the log message
     * @param key a data key
     * @param value a data value
     */
    public void error(String message, String key, Object value) {
        slf4jLogger.error(formatMessage(message, key, value));
    }

    /**
     * Log an error message with additional data.
     *
     * @param message the log message
     * @param data additional data as key-value pairs
     */
    public void error(String message, Map<String, Object> data) {
        slf4jLogger.error(formatMessage(message, data));
    }

    /**
     * Log an error message with exception.
     *
     * @param message the log message
     * @param throwable the exception
     */
    public void error(String message, Throwable throwable) {
        slf4jLogger.error(message, throwable);
    }

    /**
     * Log an error message with exception and additional data.
     *
     * @param message the log message
     * @param throwable the exception
     * @param data additional data as key-value pairs
     */
    public void error(String message, Throwable throwable, Map<String, Object> data) {
        slf4jLogger.error(formatMessage(message, data), throwable);
    }

    /**
     * Check if debug logging is enabled.
     *
     * @return true if debug logging is enabled
     */
    public boolean isDebugEnabled() {
        return slf4jLogger.isDebugEnabled();
    }

    /**
     * Check if info logging is enabled.
     *
     * @return true if info logging is enabled
     */
    public boolean isInfoEnabled() {
        return slf4jLogger.isInfoEnabled();
    }

    /**
     * Format message with key-value pair.
     */
    private String formatMessage(String message, String key, Object value) {
        return message + " [" + key + "=" + value + "]";
    }

    /**
     * Format message with data map.
     */
    private String formatMessage(String message, Map<String, Object> data) {
        if (data == null || data.isEmpty()) {
            return message;
        }

        StringBuilder sb = new StringBuilder(message);
        sb.append(" ");

        boolean first = true;
        for (Map.Entry<String, Object> entry : data.entrySet()) {
            if (!first) {
                sb.append(", ");
            }
            first = false;

            sb.append(entry.getKey())
              .append("=")
              .append(entry.getValue());
        }

        return sb.toString();
    }

    /**
     * Get a logger instance with the given name.
     *
     * <p>This is a convenience method for quickly creating a logger.
     *
     * @param name the logger name
     * @return a new Logger instance
     */
    public static Logger getLogger(String name) {
        return new Logger(name);
    }
}
